File: java/Menu.java
package ru.schoolbolt;

import java.util.ArrayList;
import java.util.Scanner;

public class Menu extends MenuItem {
    private ArrayList<MenuItem> items;
    private boolean isSubmenu;

    private static Scanner scanner = new Scanner(System.in);

    public Menu() {
        super("");
        items = new ArrayList<>();
        isSubmenu = false;
    }

    private Menu(String title) {
        super(title);
        items = new ArrayList<>();
        isSubmenu = true;
    }

    /*public void addItem(MenuItem item) {
        items.add(item);
    }*/

    public SimpleMenuItem addItem(String title, Runnable action) {
        SimpleMenuItem item = new SimpleMenuItem(title, action);
        items.add(item);
        return item;
    }

    public Menu addSubmenu(String title) {
        Menu submenu = new Menu(title);
        items.add(submenu);
        return submenu;
    }

    @Override
    public void run() {
        boolean flag = true;
        while (flag) {
            printMenu();
            flag = handleUserInput();
        }
    }

    private void printMenu() {
        for (int i = 0; i < items.size(); i++) {
            System.out.println((i + 1) + ". " + items.get(i).getTitle());
        }
        System.out.println((items.size() + 1) + (isSubmenu ? ". Выход" : ". Назад"));
    }

    private boolean handleUserInput() {
        int x = scanner.nextInt();
        x--;

        if (x == items.size()) return false;

        if (x >= 0 && x < items.size()) {
            items.get(x).run();
        } else {
            System.out.println("Нет такого пункта");
        }

        return true;
    }
}

File: Menu.cpp
#include "Menu.h"
#include <stdexcept>
Menu::Menu(std::string title) : MenuItem(title){
    isSubMenu = true;
}
Menu::Menu() : MenuItem(""){
    isSubMenu = false;
}
void Menu::printMenu(){
    std::cout << getTitle() << std::endl; 
    for (int i = 0; i < listOfMenuItem.size(); i++){
        std::cout << (i + 1) << ". " << listOfMenuItem[i] -> getTitle() << std::endl;    
    }
    if(isSubMenu){
        std::cout << listOfMenuItem.size() + 1 << ". Back" << std::endl;
    }else
        std::cout << listOfMenuItem.size() + 1 << ". Exit" << std::endl;
    
}
Menu* Menu::addsubMenu(std::string title){
    Menu* a = new Menu(title);
    listOfMenuItem.push_back(a);
    return a;
}
SimpleMenuItem* Menu::addItem(std::string title, std::function<void (void)> f){
    SimpleMenuItem* b = new SimpleMenuItem(title, f);
    listOfMenuItem.push_back(b);
    return b;
}
bool Menu::check(){
    int x;
    std::cin >> x;
    if (x == listOfMenuItem.size() + 1) return false;
    if (x > listOfMenuItem.size()){
        throw std::out_of_range("Incorrect number");
    }else if( x >= 0 && x <= listOfMenuItem.size()){
        listOfMenuItem[x-1] -> run();
    }
    return true;
}
void Menu::run(){
    bool flag = true;
    while(flag){
        printMenu();
        flag = check();
    }
}
void Menu::setStartupCommand(std::function<void (void)> command){
    startup_command = &command;
}
void Menu::setBeforeSelectCommand(std::function<void (void)> command){
    before_select_command = &command;  
} 
void Menu::setTearDownCommand(std::function<void (void)> command){
    tear_down_command = &command;
}